### PR TITLE
[New Model] Add AgiBot GO-1-Air diffusion-policy VLA (OpenPI compatible)

### PR DESCRIPTION
**Pull request by TaffyOfficial**
## Purpose

Add **AgiBot GO-1-Air** as a vLLM-Omni diffusion-policy model and keep it compatible with the OpenPI robot serving entrypoint added in #3673.

This PR covers the released `agibot-world/GO-1-Air` checkpoint path where the Latent Planner is disabled (`latent_planning: false`). The model emits a `[B, 30, 16]` action chunk through the upstream squared-cosine diffusion schedule and 5-step DPM-Solver sampling.

Refs: #3471, #3673

## What Changed

- Added `vllm_omni/diffusion/models/go1_air/` with config loading, safetensors loading, tokenizer-backed task encoding, DPM-Solver action sampling, and `Go1AirPipeline` registration.
- Added OpenPI robot serving metadata in `vllm_omni/deploy/go1_air.yaml` and default `policy_server_config` in `Go1AirConfig`.
- Added `sampling_params.extra_args["robot_obs"]` support so `/v1/realtime/robot/openpi` observations can be converted into GO-1-Air `batch_inputs`.
- Returned actions through `multimodal_output["actions"]` by making the postprocess output `{"actions": ..., "video": []}`.
- Added offline inference coverage under `tests/examples/offline_inference/test_go1_air.py` plus `examples/offline_inference/go1_air/smoke.py`.
- Added docs and supported-model entries for offline inference and OpenPI serving.

## Runtime Contract

GO-1-Air accepts either pre-built tensors in `sampling_params.extra_args["batch_inputs"]` or OpenPI-style observations in `sampling_params.extra_args["robot_obs"]`.

Required tensor schema:

| Key | Shape / type |
| --- | --- |
| `observation.state` | `torch.Tensor[B, 16]` |
| `observation.task` | `str` for `B=1`, or `list[str]` / `tuple[str, ...]` of length `B` |
| `observation.images.<camera>` | `torch.Tensor[B, history, 3, 448, 448]` |
| `observation.images.<camera>_mask` | optional scalar, `(B,)`, or `(B, history)` boolean tensor |
| `control_freq` | optional scalar or `(B,)` tensor, defaulting to 30 Hz |

OpenPI observations can provide keys such as `observation/state`, `observation/joint_position`, `observation/gripper_position`, and `observation/exterior_image_0_left`; the pipeline pads state and resizes images into the tensor schema above.

## Baseline Comparison

Single-observation synthetic benchmark on B3 (`root@106.15.124.84:31449`), `NVIDIA L20X`, `CUDA_VISIBLE_DEVICES=2`, `torch==2.11.0+cu129`, model snapshot `agibot-world/GO-1-Air@1cd2ac033fda6a85bd1610ced41007a6594e072e`.

Original baseline: `OpenDriveLab/AgiBot-World@98f136ea6b345db1a08451a2f543d67c1789a809`, `evaluate/deploy.py::GO1Infer.inference()`, `transformers==4.46.3`.

vLLM-Omni: this PR at `b3326a689e5b67a39142ce1e0ada8421e5e607ee`, `Go1AirPipeline.forward()` through `robot_obs` conversion.

Both runs used 3 warmup iterations and 30 measured iterations with one 448x448 RGB image, zero state, 30 Hz control frequency, and the same task text.

| Metric | Original GO1Infer | vLLM-Omni Go1AirPipeline | Delta |
| --- | ---: | ---: | ---: |
| Mean e2e latency | 143.642 ms | 126.670 ms | -11.8% |
| Median e2e latency | 143.887 ms | 125.417 ms | -12.8% |
| P90 e2e latency | 150.629 ms | 134.717 ms | -10.6% |
| Post-load CUDA allocated | 5.3915 GiB | 4.8286 GiB | -10.4% |
| Run peak CUDA allocated | 5.6672 GiB | 4.9480 GiB | -12.7% |
| Process RSS after run | 6.8712 GiB | 1.5938 GiB | -76.8% |
| Load time | 4.178 s | 13.116 s | slower |
| Run peak CUDA reserved | 5.7617 GiB | 10.0508 GiB | higher allocator cache |

The current implementation improves steady-state latency and allocated CUDA memory, but its safetensors loading path leaves a larger PyTorch reserved-memory cache during/after load. This is allocator-reserved cache, not live allocated tensor memory.

## Acceleration Status

This PR includes component discovery metadata and optional `torch.compile` / regional-compile hooks for GO-1-Air modules, but leaves TP/SP and quantization out of scope. TP/quantization need a separate follow-up because the action expert consumes projected VLM KV tensors and needs a real sharding plan rather than a flag-only integration.

## Test Plan

```bash
python -m py_compile \
  vllm_omni/diffusion/models/go1_air/__init__.py \
  vllm_omni/diffusion/models/go1_air/config.py \
  vllm_omni/diffusion/models/go1_air/pipeline_go1_air.py \
  examples/offline_inference/go1_air/smoke.py \
  tests/examples/offline_inference/test_go1_air.py

python -m ruff check \
  vllm_omni/diffusion/models/go1_air/__init__.py \
  vllm_omni/diffusion/models/go1_air/config.py \
  vllm_omni/diffusion/models/go1_air/pipeline_go1_air.py \
  examples/offline_inference/go1_air/smoke.py \
  tests/examples/offline_inference/test_go1_air.py

python -m ruff format --check --diff \
  vllm_omni/diffusion/models/go1_air/__init__.py \
  vllm_omni/diffusion/models/go1_air/config.py \
  vllm_omni/diffusion/models/go1_air/pipeline_go1_air.py \
  examples/offline_inference/go1_air/smoke.py \
  tests/examples/offline_inference/test_go1_air.py

git diff --check origin/main...HEAD
pytest -q tests/examples/offline_inference/test_go1_air.py
GO1_AIR_MODEL_DIR=/path/to/GO-1-Air python examples/offline_inference/go1_air/smoke.py --device cuda --dtype bfloat16
```

## Test Result

Passed locally:

```text
python -m py_compile ...
python -m ruff check ...
python -m ruff format --check --diff ...
git diff --check origin/main...HEAD
pr_scope_gate: passed
```

Passed remotely on B3 (`root@106.15.124.84:31449`, `/data/wzr/.venv`):

```text
python examples/offline_inference/go1_air/smoke.py --tiny-config --device cpu --dtype float32
[smoke] runtime_mode=no_checkpoint_policy
[smoke] action chunk=2 action_dim=4
[smoke] OK action shape=(1, 2, 4)

pytest -q tests/examples/offline_inference/test_go1_air.py
2 passed, 17 warnings in 30.97s

GO1_AIR_MODEL_DIR=<HF snapshot> python examples/offline_inference/go1_air/smoke.py --device cuda --dtype bfloat16
[smoke] runtime_mode=real_checkpoint_loaded
[smoke] action chunk=30 action_dim=16
[smoke] OK action shape=(1, 30, 16)
```

Not validated in this PR: open-loop dataset accuracy, element-wise numerical equivalence across all preprocessing variants, TP/SP, and quantization.

### CODE REVIEW

**Comment by TaffyOfficial** [comment]
## TL;DR

  **This PR (Phase 1, offline inference, LP disabled)** — register `Go1AirPipeline` and ship the model code so the
  released `agibot-world/GO-1-Air` checkpoint loads cleanly into vLLM-Omni and produces a `[B, 30, 16]` action chunk via
   5-step squared-cosine sampling on a single GPU. Verified: registry import, stub-mode pipeline, real-weight
  `load_state_dict` (0 missing / 0 unexpected), end-to-end forward with sanity-checked output stats. Includes a minimal
  smoke example, `supported_models.md` row, and a user-guide page. ~2050 LOC.

  ## Next PR (Phase 2 — open-loop evaluation)

  * `examples/offline_inference/go1_air/{end2end.py, go1_air_common.py, collect_results.sh}` mirroring the InternVLA-A1
  layout — dataset loader, deterministic noise, open-loop evaluation loop, env/GPU snapshot + result tarball.
  * `tests/examples/offline_inference/test_go1_air.py` gated on env vars, same shape as `test_internvla_a1.py`.
  * Element-wise numerical diff against `evaluate/deploy.py`'s `GO1Infer.inference()` on at least one AgiBot-World
  episode (run upstream in a separate transformers-4.46 venv and compare saved tensors).
  * Confirm-or-fix the three best-effort details flagged here: action-segment attention mask (causal vs bidirectional),
  `pixel_shuffle` v2 permute order, cross-modal RoPE alignment.
  * Image preprocessing parity with upstream's `build_transform` (InternVL mean/std).

  LP support, online serving, TP/SP/quantization, FlashAttention for the joint-sequence cross-attention, and `history >
  1` / multi-camera prompt templating remain deferred beyond Phase 2.

**Review by RuixiangMa — COMMENTED** [review]
@TaffyOfficial Suggest refactoring: consolidate sub-modules into fewer files, consistent with codebase structure

**Comment by TaffyOfficial** [comment]
Addressed the structure feedback by consolidating GO-1-Air private model internals into `model_go1_air.py`, leaving the package with `config.py`, `model_go1_air.py`, `pipeline_go1_air.py`, and `__init__.py`.

This now matches the nearby `internvla_a1` layout more closely while preserving the checkpoint-facing flat `Go1Air` module hierarchy, so safetensors keys still load without remapping.

Post-refactor local checks all passed:

```bash
python -m py_compile vllm_omni/diffusion/models/go1_air/__init__.py vllm_omni/diffusion/models/go1_air/config.py vllm_omni/diffusion/models/go1_air/model_go1_air.py vllm_omni/diffusion/models/go1_air/pipeline_go1_air.py examples/offline_inference/go1_air/smoke.py
ruff check vllm_omni/diffusion/models/go1_air examples/offline_inference/go1_air/smoke.py
ruff format --check --diff vllm_omni/diffusion/models/go1_air examples/offline_inference/go1_air/smoke.py
```

**Comment by TaffyOfficial** [comment]
Did an extra implementation pass for obvious footguns after the file consolidation.

Fixed:

- `Go1AirPolicy.forward()` now uses the model's actual parameter device/dtype for real-weight inference instead of inferring device from whichever input tensor appears first.
- State, task strings, image tensors, image masks, `control_freq`, and optional `noise` now fail early on shape/type mismatches instead of drifting into harder-to-debug attention or matmul errors.
- Image masks now support both camera-level `(B,)` masks and history-frame `(B, history)` masks, so padded history slots are dropped before `<IMG_CONTEXT>` token allocation.
- Removed the stale profiler target `policy.model.denoise_step`, which did not exist and only produced a warning when pipeline profiling was enabled.
- Replaced the `sampling_params.extra_args` fallback path with the explicit dataclass field access.

Validation after this pass:

```bash
python -m py_compile vllm_omni/diffusion/models/go1_air/__init__.py vllm_omni/diffusion/models/go1_air/config.py vllm_omni/diffusion/models/go1_air/model_go1_air.py vllm_omni/diffusion/models/go1_air/pipeline_go1_air.py examples/offline_inference/go1_air/smoke.py
ruff check vllm_omni/diffusion/models/go1_air examples/offline_inference/go1_air/smoke.py
ruff format --check --diff vllm_omni/diffusion/models/go1_air examples/offline_inference/go1_air/smoke.py
git diff --check
```

Also ran a tiny local input-validation smoke with a fake tokenizer/tiny config to cover `(B, history)` image masks, task batch mismatch, image batch mismatch, and noise shape mismatch.

**Comment by TaffyOfficial** [comment]
Ran two owner-audit passes per request: one from the GO-1-Air module-owner perspective and one from the vLLM-Omni project-owner perspective.

Addressed in commit `5cae121d3`:

- Fixed attention mask construction so valid EOS tokens are not masked when `pad_token_id == eos_token_id`.
- Real-checkpoint mode now fails fast if tokenizer loading fails; it no longer silently downgrades to zero-action stub mode.
- Strict weight loading is now the default for real checkpoints.
- Aligned several upstream-sensitive semantics: timestep embedding order (`cos, sin`), tanh GELU activations in state/action/final MLPs, `time, freq, state, action` joint-token order, and DPM-Solver sampling instead of the hand-rolled alpha-bar loop.
- Added image per-frame shape validation and component discovery metadata for offload/HSDP discovery.
- Documented the `batch_inputs` schema and fixed the docs wording from “flow-matching” to the squared-cosine diffusion schedule + DPM-Solver sampling.
- Rewrote the PR body so the Test Plan/Test Result code fences render cleanly and the latest caveats are explicit.

Latest local validation passed:

```bash
python -m py_compile vllm_omni/diffusion/models/go1_air/__init__.py vllm_omni/diffusion/models/go1_air/config.py vllm_omni/diffusion/models/go1_air/model_go1_air.py vllm_omni/diffusion/models/go1_air/pipeline_go1_air.py examples/offline_inference/go1_air/smoke.py
ruff check vllm_omni/diffusion/models/go1_air examples/offline_inference/go1_air/smoke.py
ruff format --check --diff vllm_omni/diffusion/models/go1_air examples/offline_inference/go1_air/smoke.py
git diff --check
```

Also reran the tiny-config/fake-tokenizer validation smoke through one DPM-Solver `sample_actions()` pass.

**Comment by wtomin** [comment]
@TaffyOfficial Thanks for your contributions!

I noticed that one entrypoint related to robot serving was merged #3673. Can you check if this entrypoint can support AgiBot GO-1-Air model online serving? If not possible, can you at least provide an offline inference test script for this model under `tests/examples/offline_inference`?

Another suggestion is to provide baseline comparison with original repository, including the memory consumption, e2e latency, etc.

In general, I will recommend to support some acceleration features for new models, like parallelism, quantization. Do you have plan to support at least one acceleration feature  in this PR?

**Comment by TaffyOfficial** [comment]
Addressed this in the latest update (`b3326a689e5b67a39142ce1e0ada8421e5e607ee`).

### OpenPI robot serving

GO-1-Air can now be exposed through the #3673 OpenPI robot entrypoint:

```bash
vllm serve /path/to/GO-1-Air \
  --omni \
  --deploy-config vllm_omni/deploy/go1_air.yaml
```

The PR now adds:

- `vllm_omni/deploy/go1_air.yaml` with `policy_server_config`.
- `Go1AirConfig.policy_server_config`, so the OpenPI endpoint is enabled.
- `sampling_params.extra_args["robot_obs"]` conversion in `Go1AirPipeline.forward()`.
- Action postprocess output as `{"actions": ..., "video": []}`, so the serving layer forwards actions through `multimodal_output["actions"]`.

### Offline inference test

Added `tests/examples/offline_inference/test_go1_air.py`.

Remote validation:

```text
python examples/offline_inference/go1_air/smoke.py --tiny-config --device cpu --dtype float32
[smoke] runtime_mode=no_checkpoint_policy
[smoke] action chunk=2 action_dim=4
[smoke] OK action shape=(1, 2, 4)

pytest -q tests/examples/offline_inference/test_go1_air.py
2 passed, 17 warnings in 30.97s

GO1_AIR_MODEL_DIR=<HF snapshot> python examples/offline_inference/go1_air/smoke.py --device cuda --dtype bfloat16
[smoke] runtime_mode=real_checkpoint_loaded
[smoke] action chunk=30 action_dim=16
[smoke] OK action shape=(1, 30, 16)
```

### Baseline comparison

Environment: B3 (`root@106.15.124.84:31449`), `NVIDIA L20X`, `CUDA_VISIBLE_DEVICES=2`, `torch==2.11.0+cu129`, model snapshot `agibot-world/GO-1-Air@1cd2ac033fda6a85bd1610ced41007a6594e072e`.

Original baseline: `OpenDriveLab/AgiBot-World@98f136ea6b345db1a08451a2f543d67c1789a809`, `evaluate/deploy.py::GO1Infer.inference()`, `transformers==4.46.3`.

vLLM-Omni: this PR at `b3326a689e5b67a39142ce1e0ada8421e5e607ee`, `Go1AirPipeline.forward()` through `robot_obs` conversion.

Both runs used 3 warmup iterations and 30 measured iterations with one 448x448 RGB image, zero state, 30 Hz control frequency, and the same task text.

| Metric | Original GO1Infer | vLLM-Omni Go1AirPipeline | Delta |
| --- | ---: | ---: | ---: |
| Mean e2e latency | 143.642 ms | 126.670 ms | -11.8% |
| Median e2e latency | 143.887 ms | 125.417 ms | -12.8% |
| P90 e2e latency | 150.629 ms | 134.717 ms | -10.6% |
| Post-load CUDA allocated | 5.3915 GiB | 4.8286 GiB | -10.4% |
| Run peak CUDA allocated | 5.6672 GiB | 4.9480 GiB | -12.7% |
| Process RSS after run | 6.8712 GiB | 1.5938 GiB | -76.8% |
| Load time | 4.178 s | 13.116 s | slower |
| Run peak CUDA reserved | 5.7617 GiB | 10.0508 GiB | higher allocator cache |

Steady-state latency and live allocated CUDA memory are better in this implementation. The vLLM-Omni safetensors loading path currently leaves a larger PyTorch reserved-memory cache, so I am not claiming lower reserved memory.

### Acceleration

This PR includes component discovery metadata plus optional `torch.compile` / regional-compile hooks for GO-1-Air modules. I am keeping TP/SP and quantization out of this PR because the action expert consumes projected VLM KV tensors and needs a real sharding/quantization plan rather than a flag-only integration.

File: vllm_omni/diffusion/models/go1_air/model_go1_air.py [hunk 1]
```diff
@@ -0,0 +1,1370 @@
...
+    v: torch.Tensor,
+    *,
+    num_kv_groups: int,
+    mask: torch.Tensor | None,
+    implementation: str,
+) -> torch.Tensor:
+    k_full = repeat_kv(k, num_kv_groups)
+    v_full = repeat_kv(v, num_kv_groups)
+    if implementation == "sdpa":
+        return F.scaled_dot_product_attention(q, k_full, v_full, attn_mask=mask, is_causal=mask is None)
```
Comment (wtomin) [inline vllm_omni/diffusion/models/go1_air/model_go1_air.py]: If `implementation == "sdpa"` and `mask is None`, `is_causal` will be True, and causal_mask will be applied in attention score computation.

While for `implementation == "eager"` and `mask is None`, causal_mask is never applied.

Is this mismatch between two branches intentional or accidental?

File: examples/offline_inference/go1_air/README.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
+# GO-1-Air offline inference
```
Comment (wtomin) [inline examples/offline_inference/go1_air/README.md:1]: It is a little confusing that an offline example readme file only includes a smoke test. 

I expect that this PR can provide a full inference example of this new model, as you planned in Phase 2. Is it possible to include them in this PR, and maybe remove the smoke test?

Comment (TaffyOfficial) [reply examples/offline_inference/go1_air/README.md:1]: Thanks for pointing this out. I think we should keep this PR aligned with the split in #3471: this PR is the Phase 1 model/pipeline integration plus the minimal load -> single forward -> `[B, 30, 16]` smoke coverage.

The full offline inference/open-loop example is planned for Phase 2 because it needs the representative dataset/eval harness, deterministic-noise setup, result collection, and e2e coverage to land together. So I would prefer to keep the smoke test in this PR and defer the full `end2end.py`-style example to the follow-up PR, rather than expanding the current PR scope.
